src/transport/builder/umessagebuilder.rs

```diff
@@ -0,0 +1,733 @@
+/********************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use bytes::Bytes;
+
+use crate::transport::validator::{UAttributesValidator, ValidationError};
+use crate::uprotocol::{
+    Data, UAttributes, UMessage, UMessageType, UPayload, UPayloadFormat, UPriority, UUri, UUID,
+};
+use crate::uuid::builder::UUIDBuilder;
+
+/// A builder for creating [`UMessage`]s.
+///
+/// Messages are being used by a uEntity to inform other entities about the occurrence of events
+/// and/or to invoke service operations provided by other entities.
+pub struct UMessageBuilder<'a> {
+    validator: Box<dyn UAttributesValidator>,
+    message_type: UMessageType,
+    source: Option<&'a UUri>,
+    sink: Option<&'a UUri>,
+    priority: UPriority,
+    ttl: Option<i32>,
+    token: Option<&'a String>,
+    permission_level: Option<i32>,
+    comm_status: Option<i32>,
+    request_id: Option<&'a UUID>,
+    payload: Option<Bytes>,
+    payload_format: UPayloadFormat,
+}
+
+impl<'a> Default for UMessageBuilder<'a> {
+    fn default() -> Self {
+        UMessageBuilder {
+            validator: Box::new(crate::transport::validator::PublishValidator),
```

Comment:
I was struggling with understanding how much of this we want to match the spec or be opinionated atm.

It seemed to me reading the spec for [`UTransport::send()`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#211-parameters)
> SHOULD use the Miro Uri serialization format

I realize that this is `SHOULD`, but it's at least the 'strongest' terms used for what a UUri should comply with. I see that PublishValidator [currently checks for ability to serialize down to Long format](https://github.com/SoftwareDefinedVehicle/uprotocol-sdk-rust_fork/blob/4570552d76ac8ed823dc8b68d0211f63c9229119/src/cloudevent/validator/cloudeventvalidator.rs#L355C48-L355C59), but IMHO it should do so for Micro format instead.

WDYT?

Reply:
Also tagging in @stevenhartley for some feedback, as I know, for example the Android Binder implementation _does not_ support all the numeric portions of the UUri. So if we said that the numeric portions are required only, that'd be a bit of a leak that'd let us build a message which is unable to be routed over Android Binder.

Reply:
I opened [an issue](https://github.com/eclipse-uprotocol/up-spec/issues/68) over on `up-spec` asking a few questions. I think the answer to question 1. is relevant to this PR.

Reply:
> I see that PublishValidator [currently checks for ability to serialize down to Long format](https://github.com/SoftwareDefinedVehicle/uprotocol-sdk-rust_fork/blob/4570552d76ac8ed823dc8b68d0211f63c9229119/src/cloudevent/validator/cloudeventvalidator.rs#L355C48-L355C59), but IMHO it should do so for Micro format instead.

IMHO the idea is to make sure that it is serializable to *at least* long format, which sounds reasonable to me. If a particular transport wants to use micro format, then it can still first try to serialize to micro form and could then either fall back to long format or fail sending the message altogether.

Reply:
But why _at least_ long format? IIRC the `up-client-sommr-rust` client library will probably only speak over Micro Uri format.